js/src/components/google-ads-account-card/create-account.js

```diff
@@ -2,89 +2,97 @@
+import { useState, useEffect, Fragment } from '@wordpress/element';
 
 /**
  * Internal dependencies
  */
 import Section from '.~/wcdl/section';
 import AppButton from '.~/components/app-button';
 import AccountCard, { APPEARANCE } from '.~/components/account-card';
+import useGoogleAdsAccountStatus from '.~/hooks/useGoogleAdsAccountStatus';
+import useGoogleAdsAccount from '.~/hooks/useGoogleAdsAccount';
+import useUpsertAdsAccount from '.~/hooks/useUpsertAdsAccount';
+import ClaimAccount from './claim-account';
+import ClaimAccountModal from './claim-account-modal';
 import CreateAccountButton from './create-account-button';
-import useApiFetchCallback from '.~/hooks/useApiFetchCallback';
-import { useAppDispatch } from '.~/data';
-import useDispatchCoreNotices from '.~/hooks/useDispatchCoreNotices';
+import ClaimAccountButton from './claim-account-button';
 
-const ClaimTermsAndCreateAccountButton = () => {
-	const { createNotice } = useDispatchCoreNotices();
-	const { fetchGoogleAdsAccount } = useAppDispatch();
-	const [ fetchAccountLoading, setFetchAccountLoading ] = useState( false );
-	const [ fetchCreateAdsAccount, { loading: createLoading } ] =
-		useApiFetchCallback( {
-			path: `/wc/gla/ads/accounts`,
-			method: 'POST',
-		} );
+const CreateAccount = ( props ) => {
+	const { allowShowExisting, onShowExisting } = props;
+	const [ showClaimModal, setShowClaimModal ] = useState( false );
+	const {
+		googleAdsAccount,
+		hasFinishedResolution: hasFinishedAdsAccountResolution,
+	} = useGoogleAdsAccount();
+	const {
+		hasAccess,
+		step,
+		hasFinishedResolution: hasFinishedAdsStatusResolution,
+	} = useGoogleAdsAccountStatus();
+	const [ upsertAdsAccount, { loading: isAdsCreateAccountLoading } ] =
+		useUpsertAdsAccount();
+	const shouldClaimGoogleAdsAccount = Boolean(
+		googleAdsAccount.id && hasAccess === false
+	);
 
-	const handleCreateAccount = async () => {
-		try {
-			await fetchCreateAdsAccount( { parse: false } );
-		} catch ( e ) {
-			// for status code 428, we want to allow users to continue and proceed,
-			// so we swallow the error for status code 428,
-			// and only display error message and exit this function for non-428 error.
-			if ( e.status !== 428 ) {
-				createNotice(
-					'error',
-					__(
-						'Unable to create Google Ads account. Please try again later.',
-						'google-listings-and-ads'
-					)
-				);
-				return;
-			}
-		}
+	const handleOnRequestClose = () => {
+		setShowClaimModal( false );
+	};
 
-		setFetchAccountLoading( true );
-		await fetchGoogleAdsAccount();
-		setFetchAccountLoading( false );
+	const handleOnCreateAccount = () => {
+		upsertAdsAccount();
+		setShowClaimModal( true );
 	};
 
-	return (
-		<CreateAccountButton
-			loading={ createLoading || fetchAccountLoading }
-			onCreateAccount={ handleCreateAccount }
-		/>
-	);
-};
+	const isLoading =
+		isAdsCreateAccountLoading ||
+		! ( hasFinishedAdsAccountResolution && hasFinishedAdsStatusResolution );
```

Comment:
This was very tricky. I've updated the loading property returned from `useUpsertAdsAccount` to account for both the POST request and also waiting for the Ads account and Ads Status data to be refreshed. Originally, I purposely did not do this because I was treating the refreshing of Ads data as a bit of a side effect, which would cause anything relying on that data to update, but I think returning a loading state until all three endpoints are updated makes sense.

The longer I've worked with these endpoints, it think that it might be worth considering refactoring things so that all three of these pieces of data, along with billing status, would be handled by a single API endpoint that could handle creating, updating, or reading the Ads account data. In each case, it would return data about the Ads account in a consistent shape that could be used to update the store.